// react-widgets-tailwind@1.1.5/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _plugin = _interopRequireDefault(require("tailwindcss/plugin"));

var Widget = _interopRequireWildcard(require("./widget"));

var Popup = _interopRequireWildcard(require("./popup"));

var List = _interopRequireWildcard(require("./list"));

var Combobox = _interopRequireWildcard(require("./combobox"));

var Multiselect = _interopRequireWildcard(require("./multiselect"));

var DatePicker = _interopRequireWildcard(require("./datepicker"));

var Calendar = _interopRequireWildcard(require("./calendar"));

var DropdownList = _interopRequireWildcard(require("./dropdownlist"));

var NumberPicker = _interopRequireWildcard(require("./numberpicker"));

var TimeInput = _interopRequireWildcard(require("./timeinput"));

var Listbox = _interopRequireWildcard(require("./listbox"));

const _excluded = ["addBase", "addComponents", "addUtilities"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

const isEmpty = value => {
  return value == null || value === 'null' || value === 'undefined' || value === '';
};

function removeEmpty(value) {
  if (Array.isArray(value)) {
    value.forEach(removeEmpty);
    return value;
  }

  if (value && typeof value === 'object') {
    for (let [key, v] of Object.entries(value)) {
      if (isEmpty(v)) {
        delete value[key];
      } else {
        removeEmpty(v);
      }
    }

    return value;
  }

  return value;
}

function wrap(fn, options) {
  return (decls, ...args) => {
    // console.log(removeEmpty(decls))
    return fn(removeEmpty(decls), options, ...args);
  };
}

const ComponentPlugins = {
  Popup: Popup,
  List: List,
  Listbox: Listbox,
  Autocomplete: null,
  Combobox: Combobox,
  Multiselect: Multiselect,
  DatePicker: DatePicker,
  Calendar: Calendar,
  DropdownList: DropdownList,
  NumberPicker: NumberPicker,
  TimeInput: TimeInput
};
const dependencies = {
  List: [],
  Popup: [],
  Listbox: ['List'],
  DropdownList: ['List', 'Popup'],
  Combobox: ['List', 'Popup'],
  Autocomplete: ['List', 'Popup', 'Combobox'],
  Multiselect: ['List', 'Popup'],
  DatePicker: ['Popup', 'Calendar'],
  Calendar: [],
  TimeInput: [],
  NumberPicker: []
};

function sort(components) {
  const visited = new Set();

  const topo = comp => {
    const deps = dependencies[comp] || [];
    let result = [];

    for (let dep of deps) {
      if (!visited.has(dep)) {
        result.push(...topo(dep));
      }
    }

    visited.add(comp);
    result.push(comp);
    return result;
  };

  return components.reduce((result, c) => [...result, ...topo(c)], []);
}

const defaultComponents = Object.keys(dependencies);

var _default = _plugin.default.withOptions(({
  components: _components = defaultComponents
} = {}) => {
  const enabled = sort(_components);
  return _ref => {
    let {
      addBase,
      addComponents,
      addUtilities
    } = _ref,
        args = _objectWithoutPropertiesLoose(_ref, _excluded);

    const api = Object.assign({
      addBase: wrap(addBase),
      addUtilities: wrap(addUtilities, {
        respectImportant: false
      }),
      addComponents: wrap(addComponents)
    }, args);
    Widget.plugin(api);
    enabled.forEach(comp => {
      var _ComponentPlugins$com;

      (_ComponentPlugins$com = ComponentPlugins[comp]) == null ? void 0 : _ComponentPlugins$com.plugin(api);
    });
  };
}, ({
  components: _components2 = defaultComponents
} = {}) => {
  const enabled = sort(_components2);
  return {
    theme: Object.assign({}, Widget.theme, ...enabled.map(c => {
      var _ComponentPlugins$c;

      return (_ComponentPlugins$c = ComponentPlugins[c]) == null ? void 0 : _ComponentPlugins$c.theme;
    }))
  };
});

exports.default = _default;
