// react-widgets-tailwind@1.1.5/lib/popup.js
"use strict";

exports.__esModule = true;
exports.theme = exports.plugin = void 0;
const theme = {
  rwPopup: theme => {
    const core = theme('rwCore');
    return {
      offset: theme('margin[0.5]'),
      maxHeight: theme('height.64'),
      backgroundColor: core.backgroundColor,
      border: core.borderColor,
      borderRadius: core.borderRadius,
      boxShadowOffset: '1em',
      boxShadow: theme('boxShadow.md'),
      boxShadowUp: theme('boxShadow.md'),
      zindex: 1005,
      zindexFocused: 1006
    };
  }
};
exports.theme = theme;

const plugin = ({
  theme,
  addUtilities,
  addComponents
}) => {
  const popup = theme('rwPopup');
  const core = theme('rwCore');
  addUtilities({
    ".rw-slide-transition": {
      "transition": "transform 130ms, opacity 100ms"
    },
    ".rw-slide-transition-entering": {
      "overflow": "hidden",
      "& .rw-slide-transition": {
        "transform": "translateY(0)",
        "opacity": 1,
        "transitionTimingFunction": "ease-out"
      }
    },
    ".rw-slide-transition-exiting": {
      "& .rw-slide-transition": {
        "transitionTimingFunction": "ease-in"
      }
    },
    ".rw-slide-transition-exiting,\n    .rw-slide-transition-exited": {
      "overflow": "hidden",
      "& .rw-slide-transition": {
        "opacity": 0,
        "transform": "translateY(-10%)"
      },
      "&.rw-dropup .rw-slide-transition": {
        "opacity": 0,
        "transform": "translateY(10%)"
      }
    },
    ".rw-slide-transition-exited": {
      "display": "none"
    }
  });
  addComponents({
    ".rw-popup-container": {
      "position": "absolute",
      "zIndex": popup.zindex,
      "top": "100%",
      "left": `-${popup.boxShadowOffset}`,
      "right": `-${popup.boxShadowOffset}`,
      "padding": `0 ${popup.boxShadowOffset}`,
      "&.rw-dropup": {
        "top": "auto",
        "bottom": "100%"
      },
      ".rw-state-focus &": {
        "zIndex": popup.$zindexFocused
      }
    },
    ".rw-slide-transition": {
      "marginBottom": popup.boxShadowOffset,
      ".rw-dropup > &": {
        "marginBottom": "0",
        "marginTop": popup.boxShadowOffset
      }
    },
    ".rw-popup": {
      "@apply flex flex-col": "true",
      "overflow": "auto",
      "WebkitOverflowScrolling": "touch",
      "marginTop": popup.offset,
      "borderRadius": popup.borderRadius,
      "boxShadow": popup.boxShadow,
      "border": `${popup.border} 1px solid`,
      "backgroundClip": core.backgroundClip,
      "background": popup.backgroundColor,
      "& .rw-list": {
        "maxHeight": popup.maxHeight
      },
      ".rw-dropup &": {
        "marginTop": "0",
        "marginBottom": popup.offset,
        "boxShadow": popup.boxShadowUp
      }
    }
  });
};

exports.plugin = plugin;
