// @trpc/server@10.14.0/dist/adapters/fetch/index.mjs
import { r as resolveHTTPResponse } from '../../resolveHTTPResponse-8e9dd846.mjs';
import '../../config-fb5f49a1.mjs';
import '../../TRPCError-e7333eae.mjs';
import '../../codes-52c11119.mjs';
import '../../index-972002da.mjs';
import '../../transformTRPCResponse-7a73a2df.mjs';

async function fetchRequestHandler(opts) {
    const resHeaders = new Headers();
    const createContext = async ()=>{
        return opts.createContext?.({
            req: opts.req,
            resHeaders
        });
    };
    const url = new URL(opts.req.url);
    const path = url.pathname.slice(opts.endpoint.length + 1);
    const req = {
        query: url.searchParams,
        method: opts.req.method,
        headers: Object.fromEntries(opts.req.headers),
        body: await opts.req.text()
    };
    const result = await resolveHTTPResponse({
        req,
        createContext,
        path,
        router: opts.router,
        batching: opts.batching,
        responseMeta: opts.responseMeta,
        onError (o) {
            opts?.onError?.({
                ...o,
                req: opts.req
            });
        }
    });
    for (const [key, value] of Object.entries(result.headers ?? {})){
        /* istanbul ignore if -- @preserve */ if (typeof value === 'undefined') {
            continue;
        }
        if (typeof value === 'string') {
            resHeaders.set(key, value);
            continue;
        }
        for (const v of value){
            resHeaders.append(key, v);
        }
    }
    const res = new Response(result.body, {
        status: result.status,
        headers: resHeaders
    });
    return res;
}

export { fetchRequestHandler };
