// @trpc/server@10.14.0/dist/adapters/next.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var TRPCError = require('../TRPCError-4cbab0d0.js');
var nodeHTTPRequestHandler = require('../nodeHTTPRequestHandler-7ebe1e44.js');
require('../resolveHTTPResponse-cbfdc43d.js');
require('../config-bd759464.js');
require('../codes-e0ea0f4c.js');
require('../index-4d2d31b6.js');
require('../transformTRPCResponse-d2700b72.js');

function createNextApiHandler(opts) {
    return async (req, res)=>{
        function getPath() {
            if (typeof req.query.trpc === 'string') {
                return req.query.trpc;
            }
            if (Array.isArray(req.query.trpc)) {
                return req.query.trpc.join('/');
            }
            return null;
        }
        const path = getPath();
        if (path === null) {
            const error = opts.router.getErrorShape({
                error: new TRPCError.TRPCError({
                    message: 'Query "trpc" not found - is the file named `[trpc]`.ts or `[...trpc].ts`?',
                    code: 'INTERNAL_SERVER_ERROR'
                }),
                type: 'unknown',
                ctx: undefined,
                path: undefined,
                input: undefined
            });
            res.statusCode = 500;
            res.json({
                id: -1,
                error
            });
            return;
        }
        await nodeHTTPRequestHandler.nodeHTTPRequestHandler({
            ...opts,
            req,
            res,
            path
        });
    };
}

exports.createNextApiHandler = createNextApiHandler;
