// @trpc/server@10.14.0/dist/adapters/next.mjs
import { T as TRPCError } from '../TRPCError-e7333eae.mjs';
import { n as nodeHTTPRequestHandler } from '../nodeHTTPRequestHandler-8c27f394.mjs';
import '../resolveHTTPResponse-8e9dd846.mjs';
import '../config-fb5f49a1.mjs';
import '../codes-52c11119.mjs';
import '../index-972002da.mjs';
import '../transformTRPCResponse-7a73a2df.mjs';

function createNextApiHandler(opts) {
    return async (req, res)=>{
        function getPath() {
            if (typeof req.query.trpc === 'string') {
                return req.query.trpc;
            }
            if (Array.isArray(req.query.trpc)) {
                return req.query.trpc.join('/');
            }
            return null;
        }
        const path = getPath();
        if (path === null) {
            const error = opts.router.getErrorShape({
                error: new TRPCError({
                    message: 'Query "trpc" not found - is the file named `[trpc]`.ts or `[...trpc].ts`?',
                    code: 'INTERNAL_SERVER_ERROR'
                }),
                type: 'unknown',
                ctx: undefined,
                path: undefined,
                input: undefined
            });
            res.statusCode = 500;
            res.json({
                id: -1,
                error
            });
            return;
        }
        await nodeHTTPRequestHandler({
            ...opts,
            req,
            res,
            path
        });
    };
}

export { createNextApiHandler };
